include/cyclic_reduction.hxx

```diff
@@ -282,15 +395,15 @@ public:
 	if(p == myproc) {
 	  // Just copy the data
 	  for(int i=0;i<myns; i++) {
-	    x1[sys0+i] = ifx[i][2*p];
-	    xn[sys0+i] = ifx[i][2*p+1];
+	    x1[sys0+i] = ifx(i, 2*p);
+	    xn[sys0+i] = ifx(i, 2*p+1);
 	  }
           req[p] = MPI_REQUEST_NULL;
 	}else if(nsp > 0) {
 #ifdef DIAGNOSE
           output << "Expecting receive from " << p << " of size " << len << endl;
 #endif
-	  MPI_Irecv(recvbuffer[p],
+	  MPI_Irecv(&recvbuffer(p, 0),
```

Comment:
Can we use `std::begin` here, similar to how `ifp` gets passed below? Or is that not possible due to the subscripting?

Reply:
Yes, I think the indexing prevents this from working with `std::begin`, as `recvbuffer(p,0)` refers to a single value I'm not sure what `begin` would do here.

Reply:
Most of the places I take the address or use `std::begin` with all the changes related to arrays, matrices and tensors are places I would like to change the interface of the functions to take the bounds-checked class rather than reverting to passing pointers. I've not made such changes yet as it would touch a lot of signatures and there are a couple of things that we might want to change in the design of Matrix and similar which might change the best route forward.